README.md
# OpenLayers Pdf Printer

<p align="center">
    <a href="https://www.npmjs.com/package/ol-pdf-printer">
        <img src="https://img.shields.io/npm/v/ol-pdf-printer.svg" alt="npm version">
    </a>
    <a href="https://img.shields.io/npm/dm/ol-pdf-printer">
        <img alt="npm" src="https://img.shields.io/npm/dm/ol-pdf-printer">
    </a>
    <a href="https://github.com/gastonzalba/ol-pdf-printer/blob/master/LICENSE">
        <img src="https://img.shields.io/npm/l/ol-pdf-printer.svg" alt="license">
    </a>
</p>

Customizable PDF Printer control for Open Layers. Allows the user to select the paper size, paper orientation, DPI, map elements to be displayed (compass, scale, attributions, legends, etc), add a description and more, while printing all visible layers (vector and raster). Also it allows to reframe the view before printing, using an overlay helper to better adjust to the size and aspect ratio selected.

It can export to image formats too, like png, jpeg and webp.

Tested with OpenLayers version 5, 6, 7 and 8.

## IMPORTANT

WMS sources and Icons that use external urls must be setted to `crossOrigin: 'anonymous'`. Otherwise, it will ocurr a `DOMException: Failed to execute 'toDataURL' on 'HTMLCanvasElement': Tainted canvases may not be exported`.

<div>
<img src="screenshots/example-1.jpg" style="width:100%; max-width: 650px"/>
<img src="screenshots/example-2.jpg" style="width:100%; max-width: 650px"/>
<img src="screenshots/example-3.jpg" style="width:100%; max-width: 650px"/>

## Examples

-   [Basic usage](https://raw.githack.com/GastonZalba/ol-pdf-printer/v2.1.2/examples/basic.html)

## Usage

```js
// Default options
var opt_options = {
    language: 'en',
    i18n: {...}, // Custom translations. Default is according to selected language
    showControlBtn: true,
    filename: 'Ol Pdf Printer',
    allowReframeRegionOfInterest: true,
    zoomControlOnReframe: true,
    rotationControlOnReframe: true,
    units: 'metric',
    extraInfo: {
        date: true,
        url: true,
        scale: true
    },
    mapElements: {
        description: true,
        attributions: true,
        scalebar: true,
        compass: './assets/images/compass.svg',
        legends: true
    },
    watermark: {
        title: 'Ol Pdf Printer',
        subtitle: 'https://github.com/GastonZalba/ol-pdf-printer',
        logo: false
    },
    paperSizes: [
        { size: [594, 420], value: 'A2' },
        { size: [420, 297], value: 'A3' },
        { size: [297, 210], value: 'A4', selected: true },
        { size: [210, 148], value: 'A5' }
    ],
    dpi: [
        { value: 96 },
        { value: 150, selected: true },
        { value: 200 },
        { value: 300 }
    ],
    mimeTypeExport: [
        { value: 'pdf', selected: true},
        { value: 'png' },
        { value: 'jpeg' },
        { value: 'webp' }
    ],
    style: {
        paperMargin: {
            left: 4,
            top: 4,
            right: 4,
            bottom: 4
        },
        watermark: {
            brcolor: '#000000',
            bkcolor: '#ffffff',
            txcolortitle: '#d54b4b',
            txcolorsubtitle: '#444444'
        },
        url: {
            brcolor: '#000000',
            bkcolor: '#ffffff',
            txcolor: '#0077cc'
        },
        specs: {
            brcolor: '#000000',
            bkcolor: '#ffffff',
            txcolor: '#000000'
        },
        scalebar: {
            brcolor: '#000000',
            bkcolor: '#ffffff',
            txcolor: '#000000'
        },
        attributions: {
            brcolor: '#ffffff',
            bkcolor: '#ffffff',
            txcolor: '#666666',
            txcolorlink: '#0077cc'
        },
        legends: {
            brcolor: '#000000',
            bkcolor: '#ffffff',
            txcolor: '#000000'
        },
        description: {
            brcolor: '#333333',
            bkcolor: '#333333',
            txcolor: '#ffffff'
        },
        compass: {
            brcolor: '#000000',
            bkcolor: '#333333'
        }
    },
    dateFormat: undefined, // Use browser default
    ctrlBtnClass: '',
    loader: '<span class="printLoader"></span>',
    modal: {
        animateClass: 'fade',
        animateInClass: 'show',
        transition: 300,
        backdropTransition: 0,
        templates: {
            dialog: '<div class="modal-dialog modal-dialog-centered"></div>',
            headerClose: `<button type="button" class="btn-close" data-dismiss="modal" aria-label="Close"><span aria-hidden="true">×</span></button>`
        }
    }
}

var pdfPrinter = new PdfPrinter(opt_options);

map.addControl(pdfPrinter);
```

## Changelog

See [CHANGELOG](./CHANGELOG.md) for details of changes in each release.

## Install

### Browser

#### JS

Load `ol-pdf-printer.js` after [OpenLayers](https://www.npmjs.com/package/ol), [jspdf](https://www.npmjs.com/package/jspdf) and PDF.js\[https://www.npmjs.com/package/pdfjs-dist]. Ol Pdf Printer is available as `PdfPrinter`.

```HTML
<script src="https://unpkg.com/ol-pdf-printer@2.1.2"></script>
```

#### CSS

```HTML
<link rel="stylesheet" href="https://unpkg.com/ol-pdf-printer@2.1.2/dist/css/ol-pdf-printer.css" />
<link rel="stylesheet" href="https://unpkg.com/ol-pdf-printer@2.1.2/dist/css/bootstrap.min.css" /> <!-- Bootstrap bundle -->
```

### Parcel, Webpack, etc.

NPM package: [ol-pdf-printer](https://www.npmjs.com/package/ol-pdf-printer).
Install the package via `npm`

    npm install ol-pdf-printer --save-dev

#### JS

```js
import PdfPrinter from 'ol-pdf-printer';
```

#### CSS

```js
import 'ol-pdf-printer/dist/css/ol-pdf-printer.min.css';
import 'ol-pdf-printer/dist/css/bootstrap.min.css'; // Personalized bootstrap bundle
//or scss (needs bootstrap installed)
import 'ol-pdf-printer/lib/style/scss/ol-pdf-printer.scss';
```

##### TypeScript type definition

TypeScript types are shipped with the project in the dist directory and should be automatically used in a TypeScript project. Interfaces are provided for the Options.

## API

<!-- Generated by documentation.js. Update this documentation by updating the source code. -->

#### Table of Contents

-   [PdfPrinter](#pdfprinter)
    -   [Parameters](#parameters)
    -   [showPrintSettingsModal](#showprintsettingsmodal)
    -   [hidePrintSettingsModal](#hideprintsettingsmodal)
    -   [createPdf](#createpdf)
        -   [Parameters](#parameters-1)
-   [UnitsSystem](#unitssystem)
-   [IPrintOptions](#iprintoptions)
    -   [format](#format)
    -   [orientation](#orientation)
    -   [resolution](#resolution)
    -   [scale](#scale)
    -   [regionOfInterest](#regionofinterest)
    -   [description](#description)
    -   [compass](#compass)
    -   [attributions](#attributions)
    -   [scalebar](#scalebar)
    -   [legends](#legends)
    -   [safeMargins](#safemargins)
    -   [typeExport](#typeexport)
    -   [url](#url)
    -   [date](#date)
    -   [specs](#specs)
-   [I18n](#i18n)
-   [IPaperSize](#ipapersize)
    -   [size](#size)
    -   [value](#value)
    -   [selected](#selected)
-   [IDpi](#idpi)
    -   [value](#value-1)
    -   [selected](#selected-1)
-   [brcolor](#brcolor)
-   [bkcolor](#bkcolor)
-   [txcolortitle](#txcolortitle)
-   [txcolorsubtitle](#txcolorsubtitle)
-   [brcolor](#brcolor-1)
-   [bkcolor](#bkcolor-1)
-   [txcolor](#txcolor)
-   [brcolor](#brcolor-2)
-   [bkcolor](#bkcolor-2)
-   [txcolor](#txcolor-1)
-   [txcolorlink](#txcolorlink)
-   [brcolor](#brcolor-3)
-   [bkcolor](#bkcolor-3)
-   [txcolor](#txcolor-2)
-   [brcolor](#brcolor-4)
-   [bkcolor](#bkcolor-4)
-   [txcolor](#txcolor-3)
-   [brcolor](#brcolor-5)
-   [bkcolor](#bkcolor-5)
-   [txcolor](#txcolor-4)
-   [brcolor](#brcolor-6)
-   [bkcolor](#bkcolor-6)
-   [txcolor](#txcolor-5)
-   [brcolor](#brcolor-7)
-   [bkcolor](#bkcolor-7)
-   [IStyle](#istyle)
    -   [paperMargin](#papermargin)
-   [IMimeTypeExport](#imimetypeexport)
-   [IModal](#imodal)
    -   [animateClass](#animateclass)
    -   [animateInClass](#animateinclass)
    -   [transition](#transition)
    -   [backdropTransition](#backdroptransition)
    -   [templates](#templates)
-   [IWatermark](#iwatermark)
    -   [title](#title)
    -   [subtitle](#subtitle)
    -   [logo](#logo)
-   [IExtraInfo](#iextrainfo)
    -   [date](#date-1)
    -   [url](#url-1)
    -   [specs](#specs-1)
-   [IMapElements](#imapelements)
    -   [attributions](#attributions-1)
    -   [scalebar](#scalebar-1)
    -   [compass](#compass-1)
    -   [legends](#legends-1)
-   [Options](#options)
    -   [filename](#filename)
    -   [units](#units)
    -   [allowReframeRegionOfInterest](#allowreframeregionofinterest)
    -   [zoomControlOnReframe](#zoomcontrolonreframe)
    -   [rotationControlOnReframe](#rotationcontrolonreframe)
    -   [style](#style)
    -   [extraInfo](#extrainfo)
    -   [description](#description-1)
    -   [mapElements](#mapelements)
    -   [watermark](#watermark)
    -   [paperSizes](#papersizes)
    -   [dpi](#dpi)
    -   [mimeTypeExports](#mimetypeexports)
    -   [dateFormat](#dateformat)
    -   [showControlBtn](#showcontrolbtn)
    -   [ctrlBtnClass](#ctrlbtnclass)
    -   [modal](#modal)
    -   [loader](#loader)
    -   [language](#language)
    -   [i18n](#i18n-1)

### PdfPrinter

**Extends ol/control/Control~Control**

#### Parameters

-   `opt_options` **[Options](#options)?**&#x20;

#### showPrintSettingsModal

Show the Settings Modal

Returns **void**&#x20;

#### hidePrintSettingsModal

Hide the Settings Modal

Returns **void**&#x20;

#### createPdf

Create PDF programatically without displaying the Settings Modal

##### Parameters

-   `options` **[IPrintOptions](#iprintoptions)**&#x20;
-   `showLoading` **[boolean](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Boolean)**&#x20;

Returns **void**&#x20;

### UnitsSystem

**_\[enum]_**

### IPrintOptions

**_\[interface]_**

#### format

Type: any

#### orientation

Type: (`"landscape"` | `"portrait"`)

#### resolution

Type: any

#### scale

Type: [number](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Number)

#### regionOfInterest

Area of interest. If this is provided,
the scale value is not used

Type: (Extent | Polygon)

#### description

Type: [string](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String)

#### compass

Type: [boolean](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Boolean)

#### attributions

Type: [boolean](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Boolean)

#### scalebar

Type: [boolean](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Boolean)

#### legends

Type: (LegendsOptions | [boolean](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Boolean))

#### safeMargins

Type: [boolean](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Boolean)

#### typeExport

Type: any

#### url

Type: [boolean](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Boolean)

#### date

Type: [boolean](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Boolean)

#### specs

Type: [boolean](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Boolean)

### I18n

**_\[interface]_** - Custom translations specified when creating an instance

### IPaperSize

**_\[interface]_**

#### size

Type: \[[number](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Number), [number](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Number)]

#### value

Type: [string](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String)

#### selected

Type: [boolean](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Boolean)

### IDpi

**_\[interface]_**

#### value

Type: [number](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Number)

#### selected

Type: [boolean](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Boolean)

### brcolor

Watermark border color

Type: [string](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String)

### bkcolor

Watermark background color

Type: [string](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String)

### txcolortitle

Watermark title color

Type: [string](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String)

### txcolorsubtitle

Watermark subtitle color

Type: [string](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String)

### brcolor

Url border color

Type: [string](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String)

### bkcolor

Url background color

Type: [string](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String)

### txcolor

Url text color

Type: [string](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String)

### brcolor

Attributions border color

Type: [string](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String)

### bkcolor

Attributions background color

Type: [string](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String)

### txcolor

Attributions text color

Type: [string](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String)

### txcolorlink

Attributions links color

Type: [string](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String)

### brcolor

Scalebar border color

Type: [string](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String)

### bkcolor

Scalebar background color

Type: [string](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String)

### txcolor

Scalebar text and graph color

Type: [string](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String)

### brcolor

Specs border color

Type: [string](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String)

### bkcolor

Specs background color

Type: [string](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String)

### txcolor

Specs text color

Type: [string](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String)

### brcolor

Legends border color

Type: [string](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String)

### bkcolor

Legends background color

Type: [string](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String)

### txcolor

Legends text color

Type: [string](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String)

### brcolor

Description border color

Type: [string](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String)

### bkcolor

Description background color

Type: [string](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String)

### txcolor

Description text color

Type: [string](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String)

### brcolor

Compass border color

Type: [string](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String)

### bkcolor

Compass background color

Type: [string](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String)

### IStyle

**_\[interface]_**

#### paperMargin

Only added if `Add printer margins` is checked

Type: ([number](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Number) | {top: [number](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Number), right: [number](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Number), bottom: [number](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Number), left: [number](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Number)})

### IMimeTypeExport

**_\[interface]_**

### IModal

**_\[interface]_**

#### animateClass

Type: [string](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String)

#### animateInClass

Type: [string](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String)

#### transition

Type: [number](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Number)

#### backdropTransition

Type: [number](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Number)

#### templates

Type: {dialog: ([string](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String) | [HTMLElement](https://developer.mozilla.org/docs/Web/HTML/Element))?, headerClose: ([string](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String) | [HTMLElement](https://developer.mozilla.org/docs/Web/HTML/Element))?}

### IWatermark

**_\[interface]_**

#### title

Check style section to change the color

Type: [string](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String)

#### subtitle

Check style section to change the color

Type: [string](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String)

#### logo

Display a small logo next to the title
Uns PNG format if you provide a base64 string

Type: (`false` | [string](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String) | [HTMLImageElement](https://developer.mozilla.org/docs/Web/API/HTMLImageElement) | [SVGElement](https://developer.mozilla.org/docs/Web/SVG/Element/animate))

### IExtraInfo

**_\[interface]_** - Print information at the bottom of the PDF

#### date

Print Date

Type: [boolean](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Boolean)

#### url

Current site url

Type: [boolean](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Boolean)

#### specs

DPI, Format and Scale information

Type: [boolean](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Boolean)

### IMapElements

**_\[interface]_** - MapElements

#### attributions

Layers attributions

Type: [boolean](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Boolean)

#### scalebar

Scalebar

Type: [boolean](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Boolean)

#### compass

Compass image. North must be pointing to the top

Type: (`false` | [string](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String) | [HTMLImageElement](https://developer.mozilla.org/docs/Web/API/HTMLImageElement) | [SVGElement](https://developer.mozilla.org/docs/Web/SVG/Element/animate))

#### legends

Display WMS legends

Type: (LegendsOptions | [boolean](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Boolean))

### Options

**Extends ControlOptions**

**_\[interface]_** - Options specified when creating an instance

#### filename

Export filename

Type: [string](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String)

#### units

Map unit mode

Type: [UnitsSystem](#unitssystem)

#### allowReframeRegionOfInterest

Allow to reframe a precise Region of Interest before exporting

Type: [boolean](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Boolean)

#### zoomControlOnReframe

Show zoom control when the reframe insatnce is active

Type: [boolean](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Boolean)

#### rotationControlOnReframe

Show rotation control when the reframe insatnce is active

Type: [boolean](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Boolean)

#### style

Some basic PDF style configuration

Type: [IStyle](#istyle)

#### extraInfo

Information to be inserted at the bottom of the PDF
False to disable

Type: (`false` | [IExtraInfo](#iextrainfo))

#### description

Allow add extra description to the print
False to disable

Type: [boolean](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Boolean)

#### mapElements

Elements to be showed on the PDF and in the Settings Modal.
False to disable

Type: (`false` | [IMapElements](#imapelements))

#### watermark

Watermark to be inserted in the PDF.
False to disable

Type: (`false` | [IWatermark](#iwatermark))

#### paperSizes

Paper sizes options to be shown in the settings modal

Type: [Array](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Array)<[IPaperSize](#ipapersize)>

#### dpi

DPI resolutions options to be shown in the settings modal

Type: [Array](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Array)<[IDpi](#idpi)>

#### mimeTypeExports

Export format

Type: [Array](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Array)<[IMimeTypeExport](#imimetypeexport)>

#### dateFormat

Locale time zone. Default varies according to browser locale
<https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Date/toLocaleDateString#using_locales>

Type: Locale

#### showControlBtn

Show the Btn toggler on the map

Type: [boolean](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Boolean)

#### ctrlBtnClass

ClassName to add to the Btn Control

Type: [string](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String)

#### modal

Modal configuration

Type: [IModal](#imodal)

#### loader

Element to be displayed while processing in the modal

Type: ([HTMLElement](https://developer.mozilla.org/docs/Web/HTML/Element) | [string](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String) | `false`)

#### language

Language support

Type: (`"es"` | `"en"`)

#### i18n

Add custom translations

Type: [I18n](#i18n)

## TODO

-   Interface comments
-   Add some hooks to presonalize the data before print
-   \~~Legends support~~
-   \~~Imperial units option for scalebar~~
-   \~~Customizable date format~~
-   \~~Add option to reframe the area before be captured~~
-   Events
-   Tests!


CHANGELOG.md
# Changelog

## v1.0.0
* Module created

## v1.0.1
* Fixed README details
* Removed folder "dist" and "lib" from .gitignore.

## v1.0.2
* Url links added to layer attributions

## v1.0.3
* Fixed bug on `showPrintSettingsModal` method

## v1.0.4
* Changed print button text

## v1.0.5-6
* Moved two devDependenceis to dependencies

## v1.0.7
* Added SVG support

## v1.0.8
* Added customizable `imperial` and `metric` units system
* Added customizable date format

## v1.0.9
* Added image MIME types to export
* Fixed deepObjectAssign helper (bug on merging configuration arrays)

## v1.0.10
* Updated example screenshot

## v1.0.11
* Fixed worker error on compiling lib module with webpack

## v1.0.12
* Improved worker version management and PDF.js version on example

## v1.0.13/14
* Improved typescript and rollup configuration
* Added "watch" script
* Updated dependencies
* Fixed minor css issues
* Excluded some folders from npm
* Added Source Maps

## v1.0.15
* Fixed overlapping elements on small paper prints with long urls
* Beter css builds
* Better watch mode
* Added .gitattributes
* Typescript improvements
* Changed separator on attributions

## v1.0.16
* Remove unnecessary max-width on texts

## v1.0.17
* Fix bug when printing multiples times on protrait mode

## v1.0.18
* Ol 7 compatibility
* Updated dependencies
* Updated CDNs in examples

## v1.0.19
* Minor fix on createPdf public function

## v1.0.20
* Moved modal-vanilla to peer dependencies
* Updated dependencies

## v1.0.21
* Improved README
* Added LICENSE FILE

## v1.1.0
* Removed "browser" attribute from package.json
* Added "type" module attribute to package.json
* Removed index.js
* Updated dependencies

## v1.1.1
* Improved rollup and ts configs

## v1.2.0
* Added ".js" extension on imports to work better with webpack 5 default's config
* Lib is builded with es2017 target (downgraded from esnext)
* Removed babel deps
* Added header to dist files

## v2.0.0
* Addded WMS Legends support
* Added extraInfo checkboxes
* Allow printing without safe margins (by default no margin is added)
* Improved `paperMargin` option (added customazable per margin size)
* Improved DPI accuracy
* Changed svg icon
* Fixed wrong scale transformation on exporting to image formats (now respecting the DPI)
* Improved downloading and processing status info
* Added customizable loader option
* Removed `description` attribute from inside `mapElements` option (_breaking change_)
* Removed some styles attributes from inside `watermark` option (_breaking change_)
* Added multiples customization options inside `style` attribute
* Added `showControlBtn` configuration to allow disable the control button
* Refactored using setMap method (and fixed error while using `createPdf` before the modal was displayed)
* Removed 72 DPI as a default option
* Added `createPdf` button to the example
* Replaced dependency `dom-to-image-improved` with `dom-to-image-more`
* Fixed compass not showing on firefox because of missin width and height svg attributes
* Updated to Ol8
* Updated dev dependencies
* Converted css to scss

## v2.0.1
* Added reframe ROI functionality to improve the control of the region of interest
* Removed scale parameter from the constructor and the settings modal
* Removed some forgotten console.log

## v2.0.2
* Fixed bug when legends is activated and no wms is on the map
* Fixed padding when watermark subtitle is largen than title

## v2.0.3
* Improved attributions functionality. Now extracts the data from the layers and not form the ui control
* Count loaded tiles and images in all layers (not only wms)
* Watermark logo now accepts non square images
* Added watermark logo in the example
* Added controls (zoom and rotation) on the reframe instance
* Improved css style in prinFields using flex
* Updated screenshots

## v2.0.4
* Small css fixes

## v2.0.5
* Restore contrain resolution view if reframe is cancelled

## v2.0.6
* Improved pointer events and stop event overlay
* Fixed reframe position when map is not full height and width
* Added some transparency to the reframe control buttons

## v2.1.0
* Some refactoring
* Removed `dom-to-image-more` dependency
* Improved README

## v2.1.1
* Fixed overview canvas visible in the final export (bug introduced in 2.1.0)

## v2.1.2
* Fixed bug with contraint resolution being restored before time (if it was enabled)
* Improved example

src/defaults.ts
import compassIcon from './assets/images/compass.svg';
import { I18n, Options, UnitsSystem } from './ol-pdf-printer';

export const DEFAULT_LANGUAGE = 'en';

export const defaultOptions = (i18n: I18n): Options => ({
    language: DEFAULT_LANGUAGE,
    filename: 'Ol Pdf Printer',
    allowReframeRegionOfInterest: true,
    zoomControlOnReframe: true,
    rotationControlOnReframe: true,
    style: {
        paperMargin: {
            left: 4,
            top: 4,
            right: 4,
            bottom: 4
        },
        watermark: {
            brcolor: '#000000',
            bkcolor: '#ffffff',
            txcolortitle: '#d54b4b',
            txcolorsubtitle: '#444444'
        },
        url: {
            brcolor: '#000000',
            bkcolor: '#ffffff',
            txcolor: '#0077cc'
        },
        specs: {
            brcolor: '#000000',
            bkcolor: '#ffffff',
            txcolor: '#000000'
        },
        scalebar: {
            brcolor: '#000000',
            bkcolor: '#ffffff',
            txcolor: '#000000'
        },
        attributions: {
            brcolor: '#ffffff',
            bkcolor: '#ffffff',
            txcolor: '#666666',
            txcolorlink: '#0077cc'
        },
        legends: {
            brcolor: '#000000',
            bkcolor: '#ffffff',
            txcolor: '#000000'
        },
        description: {
            brcolor: '#333333',
            bkcolor: '#333333',
            txcolor: '#ffffff'
        },
        compass: {
            brcolor: '#000000',
            bkcolor: '#333333'
        }
    },
    extraInfo: {
        date: true,
        url: true,
        specs: true
    },
    description: true,
    mapElements: {
        attributions: true,
        scalebar: true,
        legends: true,
        compass: compassIcon() as SVGElement
    },
    watermark: {
        title: 'Ol Pdf Printer',
        subtitle: 'https://github.com/GastonZalba/ol-pdf-printer',
        logo: false
    },
    paperSizes: [
        // { size: [1189, 841], value: 'A0' },
        // { size: [841, 594], value: 'A1' },
        { size: [594, 420], value: 'A2' },
        { size: [420, 297], value: 'A3' },
        { size: [297, 210], value: 'A4', selected: true },
        { size: [210, 148], value: 'A5' }
    ],
    dpi: [
        { value: 96 },
        { value: 150, selected: true },
        { value: 200 },
        { value: 300 }
    ],
    mimeTypeExports: [
        { value: 'pdf', selected: true },
        { value: 'png' },
        { value: 'jpeg' },
        { value: 'webp' }
    ],
    units: UnitsSystem.Metric,
    dateFormat: undefined,
    showControlBtn: true,
    ctrlBtnClass: '',
    loader: '<span class="printLoader"></span>',
    modal: {
        animateClass: 'fade',
        animateInClass: 'show',
        transition: 150,
        backdropTransition: 0,
        templates: {
            dialog: '<div class="modal-dialog modal-dialog-centered"></div>',
            headerClose: `<button type="button" class="btn-close" data-dismiss="modal" aria-label="${i18n.close}"><span aria-hidden="true">×</span></button>`
        }
    }
});
